File: src/doing.test.ts
import { describe, expect, test } from 'bun:test';
import { handoffPhase, isUnstarted, resolveDoingState } from './doing';
import type { AgentSession, SessionsResult } from './sessions';
import type { Handoff, Todo } from './store';

function session(overrides: Partial<AgentSession> = {}): AgentSession {
  return {
    pid: 1234,
    cwd: '/Users/x/dev/rocky-todo',
    kind: 'interactive',
    sessionId: 'sess-full-uuid',
    name: 'eelpout-a3',
    status: 'busy',
    startedAt: 0,
    ...overrides,
  };
}

function available(sessions: AgentSession[]): SessionsResult {
  return { available: true, sessions };
}

/** 세션 목록을 못 얻는 환경 — `claude` 미설치 등. */
const unavailable: SessionsResult = {
  available: false,
  sessions: [],
  reason: 'claude CLI 를 실행할 수 없다',
};

function doingTodo(
  overrides: Partial<Todo> = {},
): Pick<Todo, 'status' | 'doingBy' | 'doingSessionId'> {
  return {
    status: 'doing',
    doingBy: 'claude-code',
    doingSessionId: undefined,
    ...overrides,
  };
}

function handoff(overrides: Partial<Handoff> = {}): Handoff {
  return {
    id: 'h1',
    todoId: 't1',
    sessionId: 'sess-full-uuid',
    note: '',
    actor: 'logan',
    status: 'delivered',
    createdAt: '2026-07-30T00:00:00.000Z',
    deliveredAt: '2026-07-30T00:00:01.000Z',
    deliveredVia: 'stop',
    ...overrides,
  };
}

describe('resolveDoingState — 세션 귀속이 있는 경우', () => {
  test('세션이 살아 있고 busy 면 live', () => {
    const state = resolveDoingState(
      doingTodo({ doingSessionId: 'sess-full-uuid' }),
      'rocky-todo',
      available([session({ status: 'busy' })]),
    );
    expect(state).toBe('live');
  });

  test('세션은 있는데 idle 이면 idle — 턴이 끝났는데 done 이 안 왔다', () => {
    const state = resolveDoingState(
      doingTodo({ doingSessionId: 'sess-full-uuid' }),
      'rocky-todo',
      available([session({ status: 'idle' })]),
    );
    expect(state).toBe('idle');
  });

  test('background 세션이 state:done 이면 gone — 목록에 남아 있어도 죽은 것이다', () => {
    const state = resolveDoingState(
      doingTodo({ doingSessionId: 'sess-full-uuid' }),
      'rocky-todo',
      available([session({ kind: 'background', status: 'idle', state: 'done' })]),
    );
    expect(state).toBe('gone');
  });

  test('목록에 없으면 gone', () => {
    const state = resolveDoingState(
      doingTodo({ doingSessionId: 'sess-full-uuid' }),
      'rocky-todo',
      available([session({ sessionId: 'other' })]),
    );
    expect(state).toBe('gone');
  });

  test('spawn 이 남긴 짧은 8자 id 로도 세션을 찾는다', () => {
    // `createSpawnedHandoff` 는 full UUID 가 아니라 `claude attach` 가 받는 짧은 id 를
    // 저장한다 — sessionId 만 대조하면 살아 있는 spawn 세션이 전부 gone 으로 보인다.
    const state = resolveDoingState(
      doingTodo({ doingSessionId: 'a1b2c3d4' }),
      'rocky-todo',
      available([session({ id: 'a1b2c3d4', sessionId: 'a1b2c3d4-full-uuid', status: 'busy' })]),
    );
    expect(state).toBe('live');
  });
});

describe('resolveDoingState — 귀속이 없는 경우(보드 근사)', () => {
  test('그 보드에 세션이 하나도 없으면 gone', () => {
    const state = resolveDoingState(
      doingTodo(),
      'rocky-todo',
      available([session({ cwd: '/Users/x/dev/other-repo' })]),
    );
    expect(state).toBe('gone');
  });

  test('그 보드에 세션이 하나라도 있으면 unknown — 어느 것인지 모른다', () => {
    const state = resolveDoingState(
      doingTodo(),
      'rocky-todo',
      available([session({ cwd: '/Users/x/dev/rocky-todo' })]),
    );
    expect(state).toBe('unknown');
  });

  test('워크트리도 그 보드의 세션으로 친다 — cwd 경로 세그먼트 매칭', () => {
    const state = resolveDoingState(
      doingTodo(),
      'rocky-todo',
      available([session({ cwd: '/Users/x/dev/rocky-todo/.claude/worktrees/todo-12' })]),
    );
    expect(state).toBe('unknown');
  });

  test('사람이 잡아둔 doing 은 판정하지 않는다 — 사람은 세션 목록에 없다', () => {
    const state = resolveDoingState(
      doingTodo({ doingBy: 'logan' }),
      'rocky-todo',
      available([session({ cwd: '/Users/x/dev/other-repo' })]),
    );
    expect(state).toBe('unknown');
  });
});

describe('resolveDoingState — 판정하지 않는 경우', () => {
  test('세션 목록을 못 얻으면 무조건 unknown', () => {
    expect(
      resolveDoingState(doingTodo({ doingSessionId: 'sess-x' }), 'rocky-todo', unavailable),
    ).toBe('unknown');
    expect(resolveDoingState(doingTodo(), 'rocky-todo', unavailable)).toBe('unknown');
  });

  test('doing 이 아닌 todo 는 unknown', () => {
    const state = resolveDoingState(
      { status: 'todo', doingBy: undefined, doingSessionId: undefined },
      'rocky-todo',
      available([]),
    );
    expect(state).toBe('unknown');
  });
});

describe('handoffPhase', () => {
  test('pending / cancelled 는 status 를 그대로 쓴다', () => {
    expect(handoffPhase(handoff({ status: 'pending', deliveredAt: undefined }))).toBe('pending');
    expect(handoffPhase(handoff({ status: 'cancelled' }))).toBe('cancelled');
  });

  test('배달만 됐으면 delivered', () => {
    expect(handoffPhase(handoff())).toBe('delivered');
  });

  test('착수 기록이 있으면 accepted', () => {
    expect(handoffPhase(handoff({ acceptedAt: '2026-07-30T00:01:00.000Z' }))).toBe('accepted');
  });

  test('완료 기록이 있으면 completed', () => {
    const done = handoff({
      acceptedAt: '2026-07-30T00:01:00.000Z',
      completedAt: '2026-07-30T00:09:00.000Z',
    });
    expect(handoffPhase(done)).toBe('completed');
  });

  test('취소가 완료보다 세다 — 불가능한 조합을 표현하지 않는다', () => {
    expect(handoffPhase(handoff({ status: 'cancelled', completedAt: 'x' }))).toBe('cancelled');
  });
});

describe('isUnstarted', () => {
  test('세션이 사라졌으면 미착수다', () => {
    expect(isUnstarted(handoff(), available([session({ sessionId: 'other' })]))).toBe(true);
  });

  test('세션이 idle 이면 미착수다 — 턴이 끝났는데 아무것도 안 했다', () => {
    expect(isUnstarted(handoff(), available([session({ status: 'idle' })]))).toBe(true);
  });

  test('세션이 아직 일하는 중이면 조용하다 — 시간 임계값을 쓰지 않는다', () => {
    expect(isUnstarted(handoff(), available([session({ status: 'busy' })]))).toBe(false);
  });

  test('착수 기록이 있으면 미착수가 아니다', () => {
    const accepted = handoff({ acceptedAt: '2026-07-30T00:01:00.000Z' });
    expect(isUnstarted(accepted, available([session({ status: 'idle' })]))).toBe(false);
  });

  test('아직 배달 전(pending)이면 미착수가 아니다 — 그건 stale 이 볼 몫이다', () => {
    const pending = handoff({ status: 'pending', deliveredAt: undefined });
    expect(isUnstarted(pending, available([]))).toBe(false);
  });

  test('세션 목록을 못 얻으면 판정하지 않는다', () => {
    expect(isUnstarted(handoff(), unavailable)).toBe(false);
  });
});

File: crates/rocky-todo-core/tests/doing_test.rs
//! TS 원본 `src/doing.test.ts` 포팅.

use rocky_todo_core::doing::{
    handoff_phase, is_unstarted, resolve_doing_state, DoingState, HandoffPhase,
};
use rocky_todo_core::sessions::{AgentSession, SessionsResult};
use rocky_todo_core::types::*;

fn session() -> AgentSession {
    AgentSession {
        pid: 1234,
        cwd: "/Users/x/dev/rocky-todo".into(),
        kind: "interactive".into(),
        id: None,
        session_id: "sess-full-uuid".into(),
        name: "eelpout-a3".into(),
        status: "busy".into(),
        state: None,
        started_at: 0,
    }
}

fn available(sessions: Vec<AgentSession>) -> SessionsResult {
    SessionsResult {
        available: true,
        sessions,
        reason: None,
    }
}

fn unavailable() -> SessionsResult {
    SessionsResult::unavailable("claude CLI 를 실행할 수 없다")
}

fn doing_todo() -> Todo {
    Todo {
        id: "t1".into(),
        number: 1,
        board_id: "b1".into(),
        section_id: None,
        parent_id: None,
        title: "작업".into(),
        description: String::new(),
        status: TodoStatus::Doing,
        priority: TodoPriority::P4,
        due: None,
        labels: vec![],
        links: vec![],
        doing_by: Some("claude-code".into()),
        doing_since: None,
        doing_session_id: None,
        position: 1,
        created_at: "2026-07-01T00:00:00.000Z".into(),
        updated_at: "2026-07-01T00:00:00.000Z".into(),
        completed_at: None,
        archived_at: None,
    }
}

fn handoff() -> Handoff {
    Handoff {
        id: "h1".into(),
        todo_id: "t1".into(),
        session_id: "sess-full-uuid".into(),
        session_name: None,
        session_cwd: None,
        note: String::new(),
        actor: "logan".into(),
        status: HandoffStatus::Delivered,
        created_at: "2026-07-30T00:00:00.000Z".into(),
        delivered_at: Some("2026-07-30T00:00:01.000Z".into()),
        delivered_via: Some(HandoffVia::Stop),
        accepted_at: None,
        completed_at: None,
    }
}

// ── 세션 귀속이 있는 경우 ──

#[test]
fn attributed_busy_session_is_live() {
    let todo = Todo {
        doing_session_id: Some("sess-full-uuid".into()),
        ..doing_todo()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![session()])),
        DoingState::Live
    );
}

#[test]
fn attributed_idle_session_is_idle() {
    let todo = Todo {
        doing_session_id: Some("sess-full-uuid".into()),
        ..doing_todo()
    };
    let s = AgentSession {
        status: "idle".into(),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![s])),
        DoingState::Idle
    );
}

#[test]
fn background_state_done_is_gone_even_if_listed() {
    let todo = Todo {
        doing_session_id: Some("sess-full-uuid".into()),
        ..doing_todo()
    };
    let s = AgentSession {
        kind: "background".into(),
        status: "idle".into(),
        state: Some("done".into()),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![s])),
        DoingState::Gone
    );
}

#[test]
fn missing_session_is_gone() {
    let todo = Todo {
        doing_session_id: Some("sess-full-uuid".into()),
        ..doing_todo()
    };
    let s = AgentSession {
        session_id: "other".into(),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![s])),
        DoingState::Gone
    );
}

#[test]
fn short_spawn_id_finds_session_too() {
    // createSpawnedHandoff 는 full UUID 가 아니라 짧은 8자 id 를 저장한다.
    let todo = Todo {
        doing_session_id: Some("a1b2c3d4".into()),
        ..doing_todo()
    };
    let s = AgentSession {
        id: Some("a1b2c3d4".into()),
        session_id: "a1b2c3d4-full-uuid".into(),
        status: "busy".into(),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![s])),
        DoingState::Live
    );
}

// ── 귀속이 없는 경우(보드 근사) ──

#[test]
fn no_sessions_on_board_is_gone() {
    let s = AgentSession {
        cwd: "/Users/x/dev/other-repo".into(),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&doing_todo(), "rocky-todo", &available(vec![s])),
        DoingState::Gone
    );
}

#[test]
fn any_session_on_board_is_unknown() {
    assert_eq!(
        resolve_doing_state(&doing_todo(), "rocky-todo", &available(vec![session()])),
        DoingState::Unknown
    );
}

#[test]
fn worktree_counts_as_board_session() {
    let s = AgentSession {
        cwd: "/Users/x/dev/rocky-todo/.claude/worktrees/todo-12".into(),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&doing_todo(), "rocky-todo", &available(vec![s])),
        DoingState::Unknown
    );
}

#[test]
fn human_doing_is_not_judged() {
    let todo = Todo {
        doing_by: Some("logan".into()),
        ..doing_todo()
    };
    let s = AgentSession {
        cwd: "/Users/x/dev/other-repo".into(),
        ..session()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![s])),
        DoingState::Unknown
    );
}

// ── 판정하지 않는 경우 ──

#[test]
fn unavailable_sessions_is_always_unknown() {
    let attributed = Todo {
        doing_session_id: Some("sess-x".into()),
        ..doing_todo()
    };
    assert_eq!(
        resolve_doing_state(&attributed, "rocky-todo", &unavailable()),
        DoingState::Unknown
    );
    assert_eq!(
        resolve_doing_state(&doing_todo(), "rocky-todo", &unavailable()),
        DoingState::Unknown
    );
}

#[test]
fn non_doing_todo_is_unknown() {
    let todo = Todo {
        status: TodoStatus::Todo,
        doing_by: None,
        ..doing_todo()
    };
    assert_eq!(
        resolve_doing_state(&todo, "rocky-todo", &available(vec![])),
        DoingState::Unknown
    );
}

// ── handoffPhase ──

#[test]
fn pending_and_cancelled_use_status() {
    let pending = Handoff {
        status: HandoffStatus::Pending,
        delivered_at: None,
        ..handoff()
    };
    assert_eq!(handoff_phase(&pending), HandoffPhase::Pending);
    let cancelled = Handoff {
        status: HandoffStatus::Cancelled,
        ..handoff()
    };
    assert_eq!(handoff_phase(&cancelled), HandoffPhase::Cancelled);
}

#[test]
fn delivered_only_is_delivered() {
    assert_eq!(handoff_phase(&handoff()), HandoffPhase::Delivered);
}

#[test]
fn accepted_at_makes_accepted() {
    let h = Handoff {
        accepted_at: Some("2026-07-30T00:01:00.000Z".into()),
        ..handoff()
    };
    assert_eq!(handoff_phase(&h), HandoffPhase::Accepted);
}

#[test]
fn completed_at_makes_completed() {
    let h = Handoff {
        accepted_at: Some("2026-07-30T00:01:00.000Z".into()),
        completed_at: Some("2026-07-30T00:09:00.000Z".into()),
        ..handoff()
    };
    assert_eq!(handoff_phase(&h), HandoffPhase::Completed);
}

#[test]
fn cancelled_beats_completed() {
    let h = Handoff {
        status: HandoffStatus::Cancelled,
        completed_at: Some("x".into()),
        ..handoff()
    };
    assert_eq!(handoff_phase(&h), HandoffPhase::Cancelled);
}

// ── isUnstarted ──

#[test]
fn missing_session_is_unstarted() {
    let s = AgentSession {
        session_id: "other".into(),
        ..session()
    };
    assert!(is_unstarted(&handoff(), &available(vec![s])));
}

#[test]
fn idle_session_is_unstarted() {
    let s = AgentSession {
        status: "idle".into(),
        ..session()
    };
    assert!(is_unstarted(&handoff(), &available(vec![s])));
}

#[test]
fn busy_session_is_quiet_no_time_threshold() {
    assert!(!is_unstarted(&handoff(), &available(vec![session()])));
}

#[test]
fn accepted_is_not_unstarted() {
    let h = Handoff {
        accepted_at: Some("2026-07-30T00:01:00.000Z".into()),
        ..handoff()
    };
    let s = AgentSession {
        status: "idle".into(),
        ..session()
    };
    assert!(!is_unstarted(&h, &available(vec![s])));
}

#[test]
fn pending_is_not_unstarted() {
    let h = Handoff {
        status: HandoffStatus::Pending,
        delivered_at: None,
        ..handoff()
    };
    assert!(!is_unstarted(&h, &available(vec![])));
}

#[test]
fn unavailable_sessions_is_not_judged() {
    assert!(!is_unstarted(&handoff(), &unavailable()));
}
